hazelcast/src/main/java/com/hazelcast/internal/networking/nio/NioThread.java

```diff
@@ -393,30 +393,22 @@ public void shutdown() {
     // after we have blocked for selector.select in #runSelectLoopWithSelectorFix
     private void rebuildSelector() {
         selectorRebuildCount.inc();
-        Selector newSelector = newSelector(logger);
-        Selector oldSelector = this.selector;
+        final Selector newSelector = newSelector(logger);
+        final Selector oldSelector = this.selector;
 
-        // reset each pipeline's selectionKey, cancel the old keys
+        // each pipeline must build a new selection key
         for (SelectionKey key : oldSelector.keys()) {
             NioPipeline pipeline = (NioPipeline) key.attachment();
-
-            try {
-                int ops = key.interestOps();
-                pipeline.initSelectionKey(newSelector, ops);
-            } catch (ClosedChannelException e) {
-                logger.info("Channel was closed while trying to register with new selector.");
-            } catch (CancelledKeyException e) {
-                // a CancelledKeyException may be thrown in key.interestOps
-                // in this case, since the key is already cancelled, just do nothing
-                ignore(e);
-            }
-            key.cancel();
+            pipeline.buildNewSelectionKey(newSelector);
         }
 
-        // close the old selector and substitute with new one
-        closeSelector();
-        this.selector = newSelector;
-        logger.warning("Recreated Selector because of possible java/network stack bug.");
+        // add a task to close the selector in next select loop iteration
```

Comment:
I guess this is the fragile part because you could be closing a selector that has not yet been replaced.

You should order the older selector close after all selector replace tasks have been executed; so e.g. add an AtomicLong that is initialized with the oldSelector.keys() count and when the counter reaches zero, you close the old selector.